README.md
# NLP-Caffe

NLP-Caffe is a <a href="https://github.com/Russell91/nlp_caffe/blob/master/CHANGES.txt" target="_blank">pull request</a> [1] on the Caffe framework developed by Yangqing Jia and Evan Shelhamer, among other members of the BVLC lab at Berkeley and a large number of independent online contributers. 

This fork makes it easier for NLP users to get started without merging C++ code. The current example constructs a language model for a small subset of Google's Billion Word corpus. It uses a two-layer LSTM architecture that processes in excess of 15,000 words per second [2], and achieves a perplexity of 79. More examples for Machine Translation using the encoder-decoder model and character-level RNNs are in the works. This code will eventually be merged into the Caffe master branch. This work was funded by the <a href="http://nlp.stanford.edu/" target="_blank">Stanford NLP Group</a>, under the guidance of <a href="http://nlp.stanford.edu/~manning/" target="_blank">Chris Manning</a>, and with the invaluable expertise of <a href="http://stanford.edu/~lmthang/" target="_blank">Thang Luong</a>.

# Installation

We recommend consulting the Caffe <a href="http://caffe.berkeleyvision.org/installation.html" target="_blank">installation instructions</a> and compiling the standard Caffe library first. Next, clone this repo and run `make -j8 && make pycaffe` from the NLP-Caffe folder.

NLP-Caffe also requires <a href="https://github.com/dw/py-lmdb/" target="_blank">py-lmdb</a> at runtime, which can be installed with:

    pip install lmdb

# Example

First, cd to the caffe root directory and download the data for the model with:

    ./data/language_model/get_lm.sh

Using this data, you can generate the LMDB databases and the architecture train_val.prototxt with:

    python ./examples/language_model/create_lm.py --make_data

You'll notice this generates train, test, and validation databases in examples/language_model. It also generates the train_val.prototxt architecture file and the solver.prototxt hyperparameter file. By editing this file, you can control the hyperparameters, dataset, and architecture used by NLP-Caffe with a python interface.

You can now begin to train the network with:

    ./examples/language_model/train_lm.sh

The resulting wordvectors can be viewed with:

    ipython notebook ./examples/language_model/lm_visualization.ipynb

# Further reading

To get a better general understanding of how Caffe works, you can take advantage of the content in the <a href="http://caffe.berkeleyvision.org/installation.html" target="_blank">caffe tutorials section</a>. In particular, the <a href="http://nbviewer.ipython.org/github/BVLC/caffe/blob/master/examples/filter_visualization.ipynb" target="_blank">filter visualization tutorial</a> is a great example of how you can use the IPython notebook to investigate the weights and performance of a trained model. The <a href="http://caffe.berkeleyvision.org/gathered/examples/mnist.html" target="_blank">MNIST tutorial</a> is useful to understand how you can control Caffe over the command line and through architecture description text files.

# What's changed?

Caffe's current shared parameter implementation is pretty basic and duplicates the weight diff for each layer sharing the parameter. The net.cpp and solver.cpp files have been modified to remove this restriction and allow a constant overhead for sharing parameters that is only 50% more than the cost of a singly owned parameter, no matter how many layers are involved. Given this modification, you can build an RNN in caffe just by unrolling the number of layers you want in the prototxt file. Of course, you won't want to do this by hand, so we use the python google protobuf library to generate the prototxt. This has nice advantages like letting you create the train_val.prototxt and deploy.prototxt files simultaneously, generate random hyperparameters for multiple runs, or kick off 32 jobs at once on a cluster in just a few lines of python.

There are a few limitations of the implementation to be aware of. First, the library assumes each sentence is at most N words long, and that you have zero padded shorter sentences. The zero padding does not contribute to the loss function, but still costs processing time. There are a few ways this could be fixed, but none have been implemented yet. Fortunately, you can train on nets of a fixed max length and test on nets of a longer length. Constructing nets with a new lstm layer for each word in the sentence can also be a little daunting at first, but it's actually quite easy to manipulate all of these layers with python for loops once you get used to the idea.

An lstm layer and wordvec layer have been added, which are crucial for the language model. The slice layer has been modified to support a much faster special use case. The concat layer has received a significant speed up by taking advantage of the cublas library to perform a matrix transpose before copying memory. This allows memory to be copied in 100x bigger chunks, reducing layer forward/backward times by more than 10x when concatenating along the channel dimension. Specific changes in the code can be seen in the CHANGES.txt file.

As general advice, modifying this code to approach an entirely new NLP problem in Caffe is not for the feint of heart. On the other hand, if your research involves language models or similar problems (e.g. Part of Speech tagging), and you would like to experiment with more fancy architectures (e.g. convnet/lstm combos or clockwork lstms), then this codebase would be an excellent place to get started. You will have a very fast base implementation and all the modular goodness that comes with Caffe.

<br>

[1] All citations should be addressed to the <a href="https://github.com/BVLC/caffe" target="_blank">main Caffe repository</a>. Licensing is identical to that of a Caffe pull request.

[2] The average sentence has 22.2 words when long sentences are capped to 30 words. We train batches of 128 sentences with an average batch time of 0.166 s on a Nvidia GTX 780 TI.


CHANGES.txt
git diff d02bb2c --name-status
M	.gitignore
M	README.md
A	examples/language_model/create_lm.py
A	examples/language_model/lm_visualization.ipynb
A	examples/language_model/train_lm.sh
M	include/caffe/common_layers.hpp
M	include/caffe/net.hpp
M	include/caffe/test/test_gradient_check_util.hpp
M	include/caffe/util/math_functions.hpp
M	include/caffe/vision_layers.hpp
M	python/caffe/classifier.py
M	src/caffe/layers/concat_layer.cpp
M	src/caffe/layers/concat_layer.cu
A	src/caffe/layers/lstm_layer.cpp
A	src/caffe/layers/lstm_layer.cu
M	src/caffe/layers/slice_layer.cpp
M	src/caffe/layers/slice_layer.cu
A	src/caffe/layers/wordvec_layer.cpp
A	src/caffe/layers/wordvec_layer.cu
M	src/caffe/net.cpp
M	src/caffe/proto/caffe.proto
M	src/caffe/solver.cpp
M	src/caffe/test/test_gradient_based_solver.cpp
A	src/caffe/test/test_lstm_layer.cpp
M	src/caffe/test/test_net.cpp
M	src/caffe/test/test_slice_layer.cpp
A	src/caffe/test/test_wordvec_layer.cpp
M	src/caffe/util/math_functions.cu

git diff d02bb2c
diff --git a/.gitignore b/.gitignore
index 28f2aca..5950eed 100644
--- a/.gitignore
+++ b/.gitignore
@@ -59,12 +59,15 @@ Makefile.config
 # 1. reference, and not casually committed
 # 2. custom, and live on their own unless they're deliberated contributed
 data/*
-models/*
+#models/*
 *.caffemodel
 *.solverstate
 *.binaryproto
 *leveldb
 *lmdb
+*mdb
+*lock
+*prototxt
 
 # build, distribute, and bins (+ python proto bindings)
 build
@@ -84,6 +87,17 @@ _site
 doxygen
 docs/dev
 
+# Sublime Text settings
+*.sublime-workspace
+*.sublime-project
+
+# Eclipse Project settings
+*.*project
+
+# CMake generated files
+*.gen.cmake
+
+=======
 # LevelDB files
 *.sst
 *.ldb
diff --git a/README.md b/README.md
index ebec286..9805d34 100644
--- a/README.md
+++ b/README.md
@@ -1,34 +1,44 @@
-# Caffe
+# NLP-Caffe
 
-Caffe is a deep learning framework made with expression, speed, and modularity in mind.
-It is developed by the Berkeley Vision and Learning Center ([BVLC](http://bvlc.eecs.berkeley.edu)) and community contributors.
+NLP-Caffe is a <a href="https://github.com/Russell91/nlp_caffe/blob/master/CHANGES.txt" target="_blank">pull request</a> [1] on the Caffe framework developed by Yangqing Jia and Evan Shelhamer, among other members of the BVLC lab at Berkeley and a large number of independent online contributers. 
 
-Check out the [project site](http://caffe.berkeleyvision.org) for all the details like
+This fork makes it easier for NLP users to get started without merging C++ code. The current example constructs a language model for a small subset of Google's Billion Word corpus. It uses a two-layer LSTM architecture that processes in excess of 15,000 words per second [2], and achieves a perplexity of 79. More examples for Machine Translation using the encoder-decoder model and character-level RNNs are in the works. Hopefully, this code will eventually be merged into the Caffe master branch. This work was funded by the <a href="http://nlp.stanford.edu/" target="_blank">Stanford NLP Group</a>, under the guidance of <a href="http://nlp.stanford.edu/~manning/" target="_blank">Chris Manning</a>, and with the invaluable expertise of <a href="http://stanford.edu/~lmthang/" target="_blank">Thang Luong</a>.
 
-- [DIY Deep Learning for Vision with Caffe](https://docs.google.com/presentation/d/1UeKXVgRvvxg9OUdh_UiC5G71UMscNPlvArsWER41PsU/edit#slide=id.p)
-- [Tutorial Documentation](http://caffe.berkeleyvision.org/tutorial/)
-- [BVLC reference models](http://caffe.berkeleyvision.org/model_zoo.html) and the [community model zoo](https://github.com/BVLC/caffe/wiki/Model-Zoo)
-- [Installation instructions](http://caffe.berkeleyvision.org/installation.html)
+# Installation
 
-and step-by-step examples.
+We recommend consulting the Caffe <a href="http://caffe.berkeleyvision.org/installation.html" target="_blank">installation instructions</a> and compiling the standard Caffe library first. Next, clone this repo and run `make -j8 && make pycaffe` from the NLP-Caffe folder.
 
-[![Join the chat at https://gitter.im/BVLC/caffe](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/BVLC/caffe?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)
+NLP-Caffe also requires <a href="https://github.com/dw/py-lmdb/" target="_blank">py-lmdb</a> at runtime, which can be installed with:
 
-Please join the [caffe-users group](https://groups.google.com/forum/#!forum/caffe-users) or [gitter chat](https://gitter.im/BVLC/caffe) to ask questions and talk about methods and models.
-Framework development discussions and thorough bug reports are collected on [Issues](https://github.com/BVLC/caffe/issues).
+    pip install py-lmdb
 
-Happy brewing!
+# Tutorial
 
-## License and Citation
+First, cd to the caffe root directory and download the data for the model with:
 
-Caffe is released under the [BSD 2-Clause license](https://github.com/BVLC/caffe/blob/master/LICENSE).
-The BVLC reference models are released for unrestricted use.
+    ./data/language_model/get_lm.sh
 
-Please cite Caffe in your publications if it helps your research:
+Using this data, you can generate the LMDB databases and the architecture train_val.prototxt with:
 
-    @article{jia2014caffe,
-      Author = {Jia, Yangqing and Shelhamer, Evan and Donahue, Jeff and Karayev, Sergey and Long, Jonathan and Girshick, Ross and Guadarrama, Sergio and Darrell, Trevor},
-      Journal = {arXiv preprint arXiv:1408.5093},
-      Title = {Caffe: Convolutional Architecture for Fast Feature Embedding},
-      Year = {2014}
-    }
+    python ./examples/launguage_model/create_lm.py --make_data
+
+You'll notice this generates train, test, and validation databases in examples/ptb. It also generates the train_val.prototxt architecture file and the solver.prototxt hyperparameter file. By editing this file, you can control the hyperparameters, dataset, and architecture used by NLP-Caffe with a python interface.
+
+You can now begin to train the network with:
+

examples/language_model/create_lm.py
#!/usr/bin/env python
import sys
import lmdb
import random
import subprocess
import itertools
import argparse
import numpy as np
sys.path.append('python/caffe/proto'); import caffe_pb2

from caffe_pb2 import NetParameter, LayerParameter, DataParameter, SolverParameter, ParamSpec
from caffe_pb2 import Datum

def make_data(param):
    for phase in ['train', 'valid', 'test']:
        print 'Starting %s' % phase
        db_name = './examples/language_model/lm_%s_db' % phase
        subprocess.call(['rm', '-rf', db_name])
        env = lmdb.open(db_name, map_size=2147483648*8)

        def vocab_transform(target_input):
            def t_foo(x):
                return x if x < param['unknown_symbol'] else param['unknown_symbol']

            target_line = [t_foo(int(x)) for x in target_input.split(' ')[:param['maximum_length']]]

            target_line = target_line[:param['maximum_length']] + \
                          [param['zero_symbol']] * (param['maximum_length'] - len(target_line[:param['maximum_length']]))
            assert len(target_line) == param['maximum_length']
            return target_line

        allX = []
        with open('./data/language_model/%s_indices.txt' % phase, 'r') as f1:
            for en in f1.readlines():
                allX.append(vocab_transform(en))

        print 'Writing %s sentences' % len(allX)

        with env.begin(write=True) as txn:
            for i, target_line in enumerate(allX):
                datum = Datum()
                datum.channels = 2 * param['maximum_length']
                datum.width = 1
                datum.height = 1
                for j in range(param['maximum_length']):
                    if j == 0:
                        datum.float_data.append(param['start_symbol'])
                    else:
                        datum.float_data.append(target_line[j - 1])
                for j in range(param['maximum_length']):
                    datum.float_data.append(target_line[j])
                key = str(i)
                txn.put(key, datum.SerializeToString())

def get_solver(param):
    solver = SolverParameter()
    solver.net = param['file_train_val_net']
    solver.test_interval = param['solver_test_interval']
    solver.base_lr = param['solver_base_lr']
    solver.weight_decay = param['solver_weight_decay']
    solver.lr_policy = param['solver_lr_policy']
    solver.display = param['solver_display']
    solver.max_iter = param['solver_max_iter']
    solver.clip_gradients = param['solver_clip_gradients']
    solver.snapshot = param['solver_snapshot']
    solver.lr_policy = param['solver_lr_policy']
    solver.stepsize = param['solver_stepsize']
    solver.gamma = param['solver_gamma']
    solver.snapshot_prefix = param['solver_snapshot_prefix']
    solver.random_seed = param['solver_random_seed']
    solver.solver_mode = param['solver_solver_mode']
    solver.test_iter.append(param['solver_test_iter'])
    return solver


def get_net(param, deploy, batch_size):
    net = NetParameter()

    def add_weight_filler(param, max_value=param['init_range']):
        param.type = 'uniform'
        param.min = -max_value
        param.max = max_value

    if not deploy:
        train_data = net.layer.add()
        train_data.type = "Data"
        train_data.name = "data"
        train_data.top.append(train_data.name)
        train_data.data_param.source = 'examples/language_model/lm_train_db'
        train_data.data_param.backend = DataParameter.LMDB
        train_data.data_param.batch_size = batch_size

        test_data = net.layer.add()
        test_data.type = "Data"
        test_data.name = "data"
        test_data.top.append(test_data.name)
        test_data.data_param.source = 'examples/language_model/lm_valid_db'
        test_data.data_param.backend = DataParameter.LMDB
        test_data.data_param.batch_size = batch_size

        test_data_rule = test_data.include.add()
        test_data_rule.phase = caffe_pb2.TEST
        train_data_rule = train_data.include.add()
        train_data_rule.phase = caffe_pb2.TRAIN


    data_slice_layer = net.layer.add()
    data_slice_layer.name = "data_slice_layer"
    data_slice_layer.type = "Slice"
    data_slice_layer.slice_param.slice_dim = 1
    data_slice_layer.bottom.append('data')
    data_slice_layer.top.append('input_words')
    data_slice_layer.top.append('target_words')
    data_slice_layer.slice_param.slice_point.append(param['maximum_length'])

    label_slice_layer = net.layer.add()
    label_slice_layer.name = "label_slice_layer"
    label_slice_layer.type = "Slice"
    label_slice_layer.slice_param.slice_dim = 1
    label_slice_layer.bottom.append('target_words')
    for i in range(param['maximum_length']):
        label_slice_layer.top.append('label%d' % i)
        if i != 0:
            label_slice_layer.slice_param.slice_point.append(i)

    wordvec_layer = net.layer.add()
    wordvec_layer.name = "wordvec_layer"
    wordvec_layer.type = "Wordvec"
    wordvec_layer.bottom.append('input_words')
    wordvec_layer.top.append(wordvec_layer.name)
    wordvec_layer.wordvec_param.dimension = param['wordvec_length']
    wordvec_layer.wordvec_param.vocab_size = param['vocab_size']
    add_weight_filler(wordvec_layer.wordvec_param.weight_filler)

    wordvec_slice_layer = net.layer.add()
    wordvec_slice_layer.name = "wordvec_slice_layer"
    wordvec_slice_layer.type = "Slice"
    wordvec_slice_layer.slice_param.slice_dim = 2
    wordvec_slice_layer.bottom.append('wordvec_layer')
    for i in range(param['maximum_length']):
        wordvec_slice_layer.top.append('target_wordvec%d' % i)
        if i != 0:
            wordvec_slice_layer.slice_param.slice_point.append(i)


    for i in range(param['maximum_length']):
        if i == 0:
            dummy_layer = net.layer.add()
            dummy_layer.name = 'dummy_layer'
            dummy_layer.top.append(dummy_layer.name)
            dummy_layer.type = "DummyData"
            dummy_layer.dummy_data_param.num.append(batch_size)
            dummy_layer.dummy_data_param.channels.append(param['lstm_num_cells'])
            dummy_layer.dummy_data_param.height.append(1)
            dummy_layer.dummy_data_param.width.append(1)

            dummy_mem_cell = net.layer.add()
            dummy_mem_cell.name = 'dummy_mem_cell'
            dummy_mem_cell.top.append(dummy_mem_cell.name)
            dummy_mem_cell.type = "DummyData"
            dummy_mem_cell.dummy_data_param.num.append(batch_size)
            dummy_mem_cell.dummy_data_param.channels.append(param['lstm_num_cells'])
            dummy_mem_cell.dummy_data_param.height.append(1)
            dummy_mem_cell.dummy_data_param.width.append(1)


        for j in range(param['num_lstm_stacks']):
            concat_layer = net.layer.add()
            concat_layer.name = 'concat%d_layer%d' % (j, i)

            concat_layer.top.append(concat_layer.name)
            concat_layer.type = "Concat"
            if j == 0:
                concat_layer.bottom.append('target_wordvec%d' % i)
            if j >= 1:
                concat_layer.bottom.append('dropout%d_%d' % (j - 1, i))
            if i == 0:
                concat_layer.bottom.append(dummy_layer.name)
            else:
                concat_layer.bottom.append('lstm%d_hidden%d' % (j, i - 1))

            lstm_layer = net.layer.add()
            lstm_layer.name = 'lstm%d_layer%d' % (j, i)
            lstm_layer.type = "Lstm"
            lstm_layer.lstm_param.num_cells = param['lstm_num_cells']

            add_weight_filler(lstm_layer.lstm_param.input_weight_filler)
            add_weight_filler(lstm_layer.lstm_param.input_gate_weight_filler)
            add_weight_filler(lstm_layer.lstm_param.forget_gate_weight_filler)
            add_weight_filler(lstm_layer.lstm_param.output_gate_weight_filler)

            for k in range(4):
                param_spec = lstm_layer.param.add()
                param_spec.name = 'lstm%d_param_%d' % (j, k)
            lstm_layer.top.append('lstm%d_hidden%d' % (j, i))
            lstm_layer.top.append('lstm%d_mem_cell%d' % (j, i))
            lstm_layer.bottom.append('concat%d_layer%d' % (j, i))
            if i == 0:
                lstm_layer.bottom.append('dummy_mem_cell')
            else:
                lstm_layer.bottom.append('lstm%d_mem_cell%d' % (j, i - 1))

            dropout_layer = net.layer.add()
            dropout_layer.name = 'dropout%d_%d' % (j, i)
            dropout_layer.type = "Dropout"
            dropout_layer.top.append(dropout_layer.name)
            dropout_layer.bottom.append('lstm%d_hidden%d' % (j, i))
            dropout_layer.dropout_param.dropout_ratio = param['dropout_ratio']

    hidden_concat_layer = net.layer.add()
    hidden_concat_layer.type = "Concat"
    hidden_concat_layer.name = 'hidden_concat'
    hidden_concat_layer.top.append(hidden_concat_layer.name)
    hidden_concat_layer.concat_param.concat_dim = 0
    for i in range(param['maximum_length']):
        hidden_concat_layer.bottom.append('dropout%d_%d' % (param['num_lstm_stacks'] - 1, i))

    inner_product_layer = net.layer.add()
    inner_product_layer.name = "inner_product"
    inner_product_layer.top.append(inner_product_layer.name)
    inner_product_layer.bottom.append('hidden_concat')
    inner_product_layer.type = "InnerProduct"
    inner_product_layer.inner_product_param.bias_term = False
    inner_product_layer.inner_product_param.num_output = param['vocab_size']
    add_weight_filler(inner_product_layer.inner_product_param.weight_filler)

    label_concat_layer = net.layer.add()
    label_concat_layer.name = "label_concat"
    label_concat_layer.type = "Concat"
    label_concat_layer.concat_param.concat_dim = 0
    label_concat_layer.top.append(label_concat_layer.name)
    for i in range(param['maximum_length']):
        label_concat_layer.bottom.append('label%d' % i)

    if deploy:
        word_prob_layer = net.layer.add()
        word_prob_layer.name = "word_probs"
        word_prob_layer.top.append(word_prob_layer.name)
        word_prob_layer.type = "Softmax"
        word_prob_layer.bottom.append("inner_product")

    else:
        word_loss_layer = net.layer.add()
        word_loss_layer.name = "word_loss"
        word_loss_layer.type = "SoftmaxWithLoss"
        word_loss_layer.bottom.append("inner_product")
        word_loss_layer.bottom.append("label_concat")
        word_loss_layer.top.append(word_loss_layer.name)
        word_loss_layer.loss_param.ignore_label = param['zero_symbol']

    silence_layer = net.layer.add()
    silence_layer.name = "silence"
    silence_layer.type = "Silence"
    for j in range(param['num_lstm_stacks']):
        silence_layer.bottom.append("lstm%d_mem_cell%d" % (j, param['maximum_length'] - 1))
    for j in range(param['num_lstm_stacks'] - 1):
        silence_layer.bottom.append("dropout%d_%d" % (j, param['maximum_length'] - 1))

    return net

def write_solver(param):
    with open(param['file_solver'], 'w') as f:
        f.write(str(get_solver(param)))

def write_net(param):
    with open(param['file_train_val_net'], 'w') as f:
        f.write('name: "%s"\n' % param['net_name'])
        f.write(str(get_net(param, deploy=False, batch_size = param['train_batch_size'])))

    with open(param['file_deploy_net'], 'w') as f:
        f.write('name: "%s"\n' % param['net_name'])
        f.write('''
input: "data"
input_dim: %s
input_dim: %s
input_dim: 1
input_dim: 1

''' % (param['deploy_batch_size'], 2 * param['maximum_length']))
        f.write(str(get_net(param, deploy=True, batch_size = param['deploy_batch_size'])))


def get_base_param():
    param = {}
    param['net_name'] = "ManningNet"
    param['maximum_length'] = 30
    param['vocab_size'] = 10003
    param['num_lstm_stacks'] = 2

    param['unknown_symbol'] = param['vocab_size'] - 3
    param['start_symbol'] = param['vocab_size'] - 2
    param['zero_symbol'] = param['vocab_size'] - 1

    param['train_batch_size'] = 128
    param['deploy_batch_size'] = 128
    param['lstm_num_cells'] = 250
    param['wordvec_length'] = 250
    param['dropout_ratio'] = 0.16
    param['init_range'] = 0.14

    param['file_solver'] = "examples/language_model/solver.prototxt"
    param['file_train_val_net'] = "examples/language_model/train_val.prototxt"
    param['file_deploy_net'] = "examples/language_model/deploy.prototxt"
    param['solver_base_lr'] = 29.13
    param['solver_weight_decay'] = 1.6 * 10 ** (-6)
    param['solver_lr_policy'] = "fixed"
    param['solver_display'] = 100
    param['solver_max_iter'] = 20000
    param['solver_clip_gradients'] = 0.24
    param['solver_snapshot'] = 1000
    param['solver_lr_policy'] = 'step'
    param['solver_stepsize'] = 2500
    param['solver_gamma'] = 0.792
    param['solver_snapshot_prefix'] = "examples/language_model/lm"
    param['solver_random_seed'] = 22
    param['solver_solver_mode'] = SolverParameter.GPU
    param['solver_test_interval'] = 1000
    param['solver_test_iter'] = 200
    return param

def prepare(param):
    write_solver(param)
    write_net(param)

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--make_data', action='store_true')
    args = parser.parse_args()
    if args.make_data:
        make_data(get_base_param())
    prepare(get_base_param())

if __name__ == '__main__':
    main()
